### PR TITLE
feat: Use particle for labeling in visualizations

### PR DESCRIPTION
**Pull request by fuenfundachtzig**
* pypdt's particle() will return None for SUSY PDG IDs in 0.7.5
* backslash needs to be escaped twice
* networkx 1.11 has write_dot in submodule (no guarantees about this fix)

### CODE REVIEW

**Comment by lukasheinrich** [comment]
hi @fuenfundachtzig 

should we maybe take this opportunity to https://github.com/scikit-hep/particle. This seems to be a worthy successor for pypdt.

**Comment by fuenfundachtzig** [comment]
scikit-hep/particle is better maintained (well, just telling by the last commit dates...) -- so, yes :)

**Comment by lukasheinrich** [comment]
@fuenfundachtzig are you going to update the PR? (aside: can yuo apply `black` so the CI is passing)

Thanks!

**Comment by fuenfundachtzig** [comment]
Hi @lukasheinrich,

I have added a commit that updates to particle instead of pypdt. Unfortunately, it doesn't return BSM particles for me...?

Black is not compatible with my version of Python (3.5), so I cannot use it.

Cheers,
Alexander

**Comment by lukasheinrich** [comment]
@eduardo-rodrigues any comment on BSM support for `particle`?

@fuenfundachtzig thanks if you don't mind I'll push a formatting commit just before merging then.

**Comment by eduardo-rodrigues** [comment]
Hi @lukasheinrich, I will fill you in with details at a later stage (I realise this is advanced usage not really documented in the README) but just in case you are working on these matters overs the weekend and would appreciate a quick reply: you can do the following

```
from particle.converters.bimap import DirectionalMaps
from particle.pdgid import is_SUSY

_PDGID2PDGNameMap, _PDGName2PDGIDMap = DirectionalMaps("PDGID", "PDGNAME", converters=(int,str))
{k:v for k, v in _PDGID2PDGNameMap.items() if is_SUSY(k)}
```

and see a whole bunch of ID-name matches for BSM particles:

```
{1000001: '~d(L)',
 -1000001: '~d(L)~',
 1000002: '~u(L)',
 -1000002: '~u(L)~',
 1000003: '~s(L)',
 -1000003: '~s(L)~',
 1000004: '~c(L)',
 -1000004: '~c(L)~',
 1000005: '~b(1)',
 -1000005: '~b(1)~',
 1000006: '~t(1)',
 -1000006: '~t(1)~',
 1000011: '~e(L)-',
 -1000011: '~e(L)+',
 1000012: '~nu(eL)',
 -1000012: '~nu(eL)~',
 1000013: '~mu(L)-',
 -1000013: '~mu(L)+',
 1000014: '~nu(muL)',
 -1000014: '~nu(muL)~',
 1000015: '~tau(1)-',
 -1000015: '~tau(1)+',
 1000016: '~nu(tauL)',
 -1000016: '~nu(tauL)~',
 1000021: '~g',
 1000022: '~chi(1)0',
 1000023: '~chi(2)0',
 1000024: '~chi(1)+',
 -1000024: '~chi(1)-',
 1000025: '~chi(3)0',
 1000035: '~chi(4)0',
 1000037: '~chi(2)+',
 -1000037: '~chi(2)-',
 1000039: '~Gravitino',
 2000001: '~d(R)',
 -2000001: '~d(R)~',
 2000002: '~u(R)',
 -2000002: '~u(R)~',
 2000003: '~s(R)',
 -2000003: '~s(R)~',
 2000004: '~c(R)',
 -2000004: '~c(R)~',
 2000005: '~b(2)',
 -2000005: '~b(2)~',
 2000006: '~t(2)',
 -2000006: '~t(2)~',
 2000011: '~e(R)-',
 -2000011: '~e(R)+',
 2000012: '~nu(eR)',
 -2000012: '~nu(eR)~',
 2000013: '~mu(R)-',
 -2000013: '~mu(R)+',
 2000014: '~nu(muR)',
 -2000014: '~nu(muR)~',
 2000015: '~tau(2)-',
 -2000015: '~tau(2)+',
 2000016: '~nu(tauR)',
 -2000016: '~nu(tauR)~'}
```
It's just 4 lines of code including the imports :-).

Do get in touch if you have a list of missing BSM particles as I would be very glad to add them ...

As said, the long reply will come next - the above, and my plans, requires some explanation.

HTH.

**Comment by fuenfundachtzig** [comment]
> @fuenfundachtzig thanks if you don't mind I'll push a formatting commit just before merging then.

Yes, thanks!

**Comment by eduardo-rodrigues** [comment]
Still on the short version of the reply: I forgot to say that if you want to get LaTeX names straightway you can simple replace PDGNAME with LATEXNAME above.

**Comment by fuenfundachtzig** [comment]
I have pushed another update that uses the workaround (which currently only works for SUSY particles).

**Comment by eduardo-rodrigues** [comment]
> I have pushed another update that uses the workaround (which currently only works for SUSY particles).

The converters are not workarounds. They give you access to matches between several particle properties, and are extendible as well. What you wrote can be much simplified. I will try and commit an improvement by tomorrow, so do not merge just now.

**Comment by lukasheinrich** [comment]
hi @cburton12 -- that sounds great actually.. does it incur any new dependencies?

**Comment by cburton12** [comment]
In principle, this should only use pygraphviz, which is already a dependency of networkx. However, I do remember some oddity where I had to install graphvix _and_ pygraphviz separately.

But, that may have been one of those weird Anaconda+PyPI things. (I've been with conda so long now, I've forgotten how pip works...)

I just uploaded the code to a public place in case you just want to copy/paste it to this MR. The output of the main script is attached. 
image

**Comment by lukasheinrich** [comment]
@cburton12 this seems nice. @fuenfundachtzig any thoughts?

**Comment by fuenfundachtzig** [comment]
Indeed, graphviz and pygraphviz have to be installed additionally (plus matplotlib), but the workflow may be simpler.

**Comment by eduardo-rodrigues** [comment]
Ha, this is a bit what I did in `DecayLanguage` for particle decays (typically in decay files), see the plot in https://github.com/scikit-hep/decaylanguage/blob/master/README.md :-).

File: src/pylhe/__init__.py
Comment (eduardo-rodrigues) [inline src/pylhe/__init__.py:237]: This mapping should be done once and for all, so please remove it from the loop ;-). And it is probably good to `del` these at the end of the file ...

File: src/pylhe/__init__.py
Comment (eduardo-rodrigues) [inline src/pylhe/__init__.py:235]: I gave you this workaround for SUSY particles which I work on a extensive data file with "all" particles. This CSV file you access allows one to get mapping between several particle properties without the need to create all `Particle` instances. The mapping works for all particles, so no need to do anything special for SUSY. You can trivially look at the contents of `_PDGID2PDGNameMap` and you will see what I mean …

In other words, no need for the `else`.

File: src/pylhe/__init__.py
Comment (eduardo-rodrigues) [reply src/pylhe/__init__.py:237]: The mapping gives you LaTeX names straightaway so I would rather define `_PDGID2LaTeXNameMap, _LaTeXName2PDGIDMap`.

**Comment by eduardo-rodrigues** [comment]
Folks, anything else you would need from me to sort the PR out, resolve the conflicts and merge? I think you have a lot of nice updates pipelined for a new major release … ;-)

**Comment by eduardo-rodrigues** [comment]
@fuenfundachtzig, @lukasheinrich, no reaction to the comments and change suggestions?

**Comment by eduardo-rodrigues** [comment]
It's been 2 months @lukasheinrich and @fuenfundachtzig since I made some suggestions ... No (re)action?

**Comment by lukasheinrich** [comment]
@eduardo-rodrigues maybe it's worth opening a separate PR on this? I was waiting to see if @fuenfundachtzig wants to incorporarte @cburton12's suggestion.

**Comment by eduardo-rodrigues** [comment]
I'm happy to help in a separate PR, sure. On the other hand my 2 suggestions above are really easy hence can trivially be done here. I will see how this PR evolves ... :-). Thanks.

**Comment by cburton12** [comment]
Feel free to leave my suggestion out. I could make a separate MR once this one is picked up.

**Comment by lukasheinrich** [comment]
ok then I would suggest @fuenfundachtzig should just go ahead w/ the base PR and we can implement @eduardo-rodrigues and @cburton12's suggestion in follow-up PRs.

**Comment by matthewfeickert** [comment]
Hi @fuenfundachtzig. I just wanted to touch base here as @eduardo-rodrigues reminded me about this PR and see what the state of things is. There's been some delay on our part (sorry) so do you have time at the moment to rebase this and revisit the state of it with us? Or should @lukasheinrich and I just take care of it when we get around to it?

**Comment by fuenfundachtzig** [comment]
Hi all,

I have integrated and tested the suggested changes by eduardo-rodrigues.

Cheers
Alexander

**Comment by matthewfeickert** [comment]
> I have integrated and tested the suggested changes

Thanks @fuenfundachtzig. There's conflicts between this branch and `master` so this should get rebased to address them. Do you want to take care of this or would you prefer that @lukasheinrich and I do it?

File: src/pylhe/__init__.py
Comment (matthewfeickert) [inline src/pylhe/__init__.py:175]: We should have an extra space between the `:` and the `"` so you don't have text smashed up against each other. 
```suggestion
    except ET.ParseError as e:
        print("WARNING. Parse Error: ", e)
```

File: src/pylhe/__init__.py
Comment (matthewfeickert) [inline src/pylhe/__init__.py:228]: It would be great if this could be an actual docstring instead of just a comment.

File: src/pylhe/__init__.py
Comment (matthewfeickert) [inline src/pylhe/__init__.py:254]:
```suggestion
```
Whitespace changes should get removed.

**Review by matthewfeickert — CHANGES_REQUESTED** [review]
Looking good so far. Just a few comments.

**Comment by matthewfeickert** [comment]
> I just uploaded the code to a public place in case you just want to copy/paste it to this MR. https://gitlab.cern.ch/cburton/JpsiVanalysis/-/blob/master/MCProduction/eventGrapher.py#L48

@cburton12 This URL is no longer valid. Can you link us to the new version of the example code (if it still exists) using a permalink?

**Comment by cburton12** [comment]
I think I just made a typo in the previous link. The correct address has a capital E: https://gitlab.cern.ch/cburton/JpsiVanalysis/-/blob/master/MCProduction/EventGrapher.py
Let me know if this is ok.

**Comment by matthewfeickert** [comment]
> Let me know if this is ok.

Thanks! That was it. :)

File: src/pylhe/__init__.py
Comment (fuenfundachtzig) [reply src/pylhe/__init__.py:175]: The comma inserts a blank.

File: src/pylhe/__init__.py
Comment (fuenfundachtzig) [reply src/pylhe/__init__.py:228]: But this comment does not explain what the function is doing?

**Comment by fuenfundachtzig** [comment]
> Thanks @fuenfundachtzig. There's conflicts between this branch and `master` so this should get rebased to address them. Do you want to take care of this or would you prefer that @lukasheinrich and I do it?

The structure of `setup.py` has changed and I don't know where to put the `install_requires` now?

**Comment by eduardo-rodrigues** [comment]
Hi all, I will have a look at the updates concerning `Particle` later today ... Seems we're almost ready to merge.

File: src/pylhe/__init__.py
Comment (eduardo-rodrigues) [inline src/pylhe/__init__.py:24]: What is this duplication of `visualize`?

**Comment by eduardo-rodrigues** [comment]
> > Thanks @fuenfundachtzig. There's conflicts between this branch and `master` so this should get rebased to address them. Do you want to take care of this or would you prefer that @lukasheinrich and I do it?
> 
> The structure of `setup.py` has changed and I don't know where to put the `install_requires` now?

Those now go in `setup.cfg`  ...

File: src/pylhe/__init__.py
Comment (eduardo-rodrigues) [reply src/pylhe/__init__.py:24]: Otherwise all looks good, modulo the change pypdt to particle in setup.cfg.

File: src/pylhe/__init__.py
Comment (matthewfeickert) [reply src/pylhe/__init__.py:228]: Comments and docstrings are not actually the same thing. Check out [PEP 257](https://www.python.org/dev/peps/pep-0257/) for some of the differences, but the main thing that matters here is that if you ran

```python
>>> import pylhe
>>> help(pylhe.visualize)
```
you would just get
```
visualize(event, outputname)
```

and no information as the comment gets stripped out. Docstrings don't and so are actually useful documentation.

File: src/pylhe/__init__.py
Comment (matthewfeickert) [reply src/pylhe/__init__.py:24]: > What is this duplication of `visualize`?

@fuenfundachtzig can verify, but I assume it is because at the moment all the code is in this single `__init__.py` file that to give the `LHEfile` a `visualize` method they didn't need to import `visualize`.

This is a bit awkward looking, but assuming I'm not missing something else this could get cleaned up more down the road with some refactoring of the library structure.

**Comment by matthewfeickert** [comment]
Hm, I messed something up when I was pushing some of the fixes. These are properly under `feat/add-particle-labeling` now, but when I tried to push them to @fuenfundachtzig's fork this PR got closed. Nothing was lost though, so we can fix this.

**Comment by matthewfeickert** [comment]
Oh, it seems that I can't fix this as I don't have owner permissions like I thought I did &mdash; only @lukasheinrich does. So when I tried to push to @fuenfundachtzig's fork things got messed up. This can get fixed in 2 ways:

1. @lukasheinrich adds @fuenfundachtzig's remote locally and then does a force push of `feat/add-particle-labeling` to @fuenfundachtzig's `master` branch (as that's the source branch this PR started on).
2. @fuenfundachtzig does the following:
```
$ git fetch --all
$ git checkout upstream/feat/add-particle-labeling -b feat/add-particle-labeling # Assuming the remote is named upstream
$ git checkout master
$ git merge feat/add-particle-labeling 
```

...or just
```
$ git fetch --all
$ git checkout master
$ git reset --hard upstream/feat/add-particle-labeling
```

After that this PR can get reopened and it should be like nothing happened. Sorry about this. :(

**Comment by matthewfeickert** [comment]
I can of course also just open up a PR from `feat/add-particle-labeling` and make sure that the author is set to @fuenfundachtzig so they get the proper credit in the contribution record if that would be easier, but it would be cleaner from a continuity perspective if this PR could just get reopened as described above.

**Comment by eduardo-rodrigues** [comment]
Strange, @matthewfeickert, I thought any org admin could in principle reopen PRs on repos. Let's then wait for @lukasheinrich ... :-)

**Comment by matthewfeickert** [comment]
> Strange, @matthewfeickert, I thought any org admin could in principle reopen PRs on repos.

So my original message which I since deleted was wrong. It isn't an issue of permissions, it is that in GitHub's view there is nothing to reopen as there is no diff (note how the "Reopen and comment" button at the bottom is grayed out). I added the following to my original post later on:

> (I think this happened because I did a rebase and as this PR started on `fuenfundachtzig:master` instead of a feature branch when the state of `scikit-hep:master` hit `fuenfundachtzig:master` GitHub thought "the state of the source branch is the same as the target branch. There is nothing to diff so I'm going to close this PR" and did so before the rest of the rebase could finish.)

So we actually need @fuenfundachtzig to follow the instructions I added above to reopen this, or for me to open a new PR but give them credit on it.

**Comment by eduardo-rodrigues** [comment]
Thanks @matthewfeickert for the clarification.

File: src/pylhe/__init__.py
Comment (fuenfundachtzig) [reply src/pylhe/__init__.py]: Right, what is mean is that a proper docstring would describe the function as a whole, whereas the comment you're refering to just explains the one statement that follows it.

File: src/pylhe/__init__.py
Comment (fuenfundachtzig) [reply src/pylhe/__init__.py]: Correct.

**Comment by fuenfundachtzig** [comment]
> @fuenfundachtzig does:
> ```
> $ git fetch --all
> $ git checkout master
> $ git reset --hard upstream/feat/add-particle-labeling
> $ git push origin master
> ```
> 
> After that this PR can get reopened and it should be like nothing happened. I can't reopen it now as there are no commits different between the project `master` and the fork `master`. Sorry about this. :(

I did that but cannot reopen this PR. I can make a new PR though.

File: src/pylhe/__init__.py
Comment (fuenfundachtzig) [reply src/pylhe/__init__.py]: (I also added a docstring.)

**Comment by matthewfeickert** [comment]
> I did that but cannot reopen this PR. I can make a new PR though.

@fuenfundachtzig I was able to reopen it. Thanks!

**Review by eduardo-rodrigues — APPROVED** [review]

File: src/pylhe/__init__.py
Comment (matthewfeickert) [inline src/pylhe/__init__.py:177]: We're currently failing CI here given `flake8`'s [E722 do not use bare 'except'](https://www.flake8rules.com/rules/E722.html) but this isn't the fault of this PR. I'll fix this here as soon as I'm out of my next meeting.

**Review by matthewfeickert — CHANGES_REQUESTED** [review]
I'll fix this up before this gets merged as it isn't @fuenfundachtzig's fault but just something that needed to get done.

File: src/pylhe/__init__.py
Comment (matthewfeickert) [reply src/pylhe/__init__.py:177]: Actually this isn't here, but lower down.

File: src/pylhe/__init__.py
Comment (matthewfeickert) [inline src/pylhe/__init__.py:251]: We're currently failing CI here given `flake8`'s [E722 do not use bare 'except'](https://www.flake8rules.com/rules/E722.html) I'll fix this here as soon as I'm out of my next meeting.

**Review by matthewfeickert — APPROVED** [review]
Okay. Thank you very much for this nice PR @fuenfundachtzig! We really appreciate your contribution and thank you for your patience. :)

**Pull request merged**
